File: KNN/KNN1/KNN1.py
from numpy import *
import matplotlib.pyplot as plt


spt=700
size=1000
k=20

mat=[]
label=[]

def readData():
    filename="KNN/KNN1/data/dataset.txt"
    f=open(filename,"r",encoding='utf-8')
    data=f.readlines()
    lines=len(data)
    returnMat=zeros((lines+1,3))
    label=[0]
    index=1
    for line in data:
        line=line.strip()
        inform=line.split("\t")
        returnMat[index]=inform[:3]
        label+=[inform[-1]]
        index+=1
    f.close()
    return returnMat,label

def dis(x,y):
    return sqrt(sum((x-y)**2))

def test(x):
    tlist=[0.0]
    for i in range(1,spt+1):
        tlist+=[dis(mat[i],mat[x])]
    tlist.sort()
    ans=[0,0,0,0]
    for i in range(1,k+1):
        ans[label[i]]+=1
    return max(ans[1:])


if __name__ == '__main__':
    mat,label=readData()
    label = list(map(int,label))
    fig = plt.figure()
    ax = plt.axes(projection='3d')
    ax.scatter(mat[1:,0],mat[1:,1],mat[1:,2],15.0*array(label[1:]),15.0*array(label[1:]),15.0*array(label[1:]))
    plt.show()
    crt=0
    for i in range(spt-1,1000):
        if test(i)==label[i]:
            crt+=1
    print(crt/(1001-spt))

File: KNN/KNN1/KNN1.cpp
#include<bits/stdc++.h>
#include "matplotlibcpp.h"
namespace plt = matplotlibcpp;
using namespace std;
const int size_a=3,N=1000,size_train=700,type=3,showId_x=1,showId_y=2;
int k=20;
double maxa[size_a+1],mina[size_a+1];
struct node{
    double a[size_a+1];//属性
    int id;//标签
}d[N+1];
void readData(){
    //输入
    freopen("data\\dataset.txt","r",stdin);
    for(int i=1;i<=N;i++){
        for(int j=1;j<=size_a;j++){
            cin>>d[i].a[j];
            if(d[i].a[j]>maxa[j]){
                maxa[j]=d[i].a[j];
            }
            if(d[i].a[j]<mina[j]){
                mina[j]=d[i].a[j];
            }
        }
        cin>>d[i].id;
    }
    //线性归一化
    vector<double> x,y;
    for(int i=1;i<=N;i++){
        for(int j=1;j<=size_a;j++){
            d[i].a[j]=(d[i].a[j]-mina[j])/(maxa[j]-mina[j]);
        }
        x.push_back(d[i].a[showId_x]);
        y.push_back(d[i].a[showId_y]);
    }
    //绘图
    plt::scatter(x,y);
}
double dis(int x,int y){
    double ans=0;
    for(int i=1;i<=size_a;i++){
        ans+=(d[x].a[i]-d[y].a[i])*(d[x].a[i]-d[y].a[i]);
    }
    return sqrt(ans);
}
int testOne(int x){
    double temp[size_train+1]={2e9};
    int ans[size_train+1]={};
    for(int i=1;i<=size_train;i++){
        temp[i]=dis(x,i);
        for(int j=1;j<=k;j++){
            if(temp[ans[j]]>temp[i]){
                for(int l=k;l>j;l--){
                    ans[l]=ans[l-1];
                }
                ans[j]=i;
                break;
            }
        }
    }
    int cnt[type+1]={},rtn=0,maxnum=-1;
    for(int i=1;i<=k;i++){
        cnt[d[ans[i]].id]++;
    }
    for(int i=1;i<=type;i++){
        if(cnt[i]>maxnum){
            maxnum=cnt[i];
            rtn=i;
        }
    }
    vector<double> u,v;
    for(int i=1;i<=k;i++){
        u.push_back(d[ans[i]].a[showId_x]);
        v.push_back(d[ans[i]].a[showId_y]);
    }
    plt::scatter(u,v);
    return rtn;
}
double testAll(){
    int cnt=0;
    for(int i=size_train+1;i<=N;i++){
        if(testOne(i)==d[i].id){
            cnt++;
        }
    }
    return cnt*1.0/(N-size_train);
}
int main(){
    readData();
    /*//k对成功率的影响
    vector<double> x,ans;
    for(int i=1;i<=100;i++){
        k=i;
        x.push_back(i);
        ans.push_back(testAll());
    }
    plt::plot(x,ans);
    plt::show();*/
    cout<<testAll();
    plt::xlim(0,1);
    plt::ylim(0,1);
    plt::show();
    return 0;
}